### PR TITLE
Gui and algoritms to check

### PR DESCRIPTION
**Pull request by RadekBucki**

### CODE REVIEW

File: src/main/java/pl/numerki/backend/Bisection.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+import static java.lang.Math.abs;
+
+public class Bisection {
+    public static double getZeroPositionsWithPrecisionCondition(
+            Function<Double, Double> function,
+            double leftEndOfCompartment,
+            double rightEndOfCompartment,
+            double epsilon
+    ) throws Exception {
```
Comment (kambed) [inline src/main/java/pl/numerki/backend/Bisection.java]: should be BisectionException

File: src/main/java/pl/numerki/frontend/MainFormController.java [hunk 2]
```diff
@@ -1,18 +1,169 @@
...
+        }
+        return switch (chosenFunction) {
+            case "ax^2+bx+c" -> squareFunction(a, b, c);
+            case "sin(x)" -> sinFunction();
+            case "a^x" -> exponentialFunction(a);
+            case "sin(ax^2+bx+c)" -> complexFunction(a, b, c);
+        };
+    }
+
+    public void squareFunctionChosen(ActionEvent actionEvent) {
```
Comment (kambed) [inline src/main/java/pl/numerki/frontend/MainFormController.java]: do function with 3 bool args to set visibility of a b c in *Chosen functions

File: src/main/java/pl/numerki/frontend/MainFormController.java [hunk 3]
```diff
@@ -1,18 +1,169 @@
...
+        b.setVisible(true);
+        c.setVisible(true);
+    }
+
+    private void setChosenFunction(ActionEvent actionEvent) {
+        function.setText(((MenuItem) actionEvent.getSource()).getText());
+        chosenFunction = ((MenuItem) actionEvent.getSource()).getText();
+    }
+
+    public void showIterations(ActionEvent actionEvent) {
```
Comment (kambed) [inline src/main/java/pl/numerki/frontend/MainFormController.java]: do function with 2 bool values for precision and iterations as well

File: src/main/java/pl/numerki/frontend/MainFormController.java [hunk 4]
```diff
@@ -1,18 +1,169 @@
...
+    private Function<Double, Double> getFunction() throws Exception {
+        double a, b, c;
+        try {
+            a = Double.parseDouble(this.a.getText());
+            b = Double.parseDouble(this.b.getText());
+            c = Double.parseDouble(this.c.getText());
+        } catch (NumberFormatException e) {
+            throw new Exception("Nie podano liczb");
+        }
+        return switch (chosenFunction) {
```
Comment (kambed) [inline src/main/java/pl/numerki/frontend/MainFormController.java]: NO default value compilation error

File: src/main/java/pl/numerki/frontend/MainFormController.java [hunk 5]
```diff
@@ -1,18 +1,169 @@
...
+                int iterations = Integer.parseInt(this.iterations.getText());
+                resultBisection = Bisection.getZeroPositionsWithIterationCondition(function, from, to, iterations);
+                resultSecant = SecantMethod.getZeroPositionsWithIterationCondition(function, from, to, iterations);
+            }
+        } catch (BisectionException e) {
+            bisectionException = true;
+        } catch (SecantException e) {
+            secantException = true;
+        } catch (Exception e) {
+            AlertBox.alertShow(
```
Comment (kambed) [inline src/main/java/pl/numerki/frontend/MainFormController.java]: Shouldn't there be some error or sth like that?

File: src/main/java/pl/numerki/backend/Bisection.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+        }
+        throw new BisectionException();
+    }
+
+    public static double getZeroPositionsWithIterationCondition(
+            Function<Double, Double> function,
+            double leftEndOfCompartment,
+            double rightEndOfCompartment,
+            int numberOfIterations
+    ) throws Exception {
```
Comment (kambed) [inline src/main/java/pl/numerki/backend/Bisection.java]: same thing Bisextion excpetion

Comment (kambed) [inline src/main/java/pl/numerki/backend/Bisection.java]: same thing Bisextion excpetion

File: src/main/java/pl/numerki/backend/SecantMethod.java [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+import static java.lang.Math.abs;
+
+public class SecantMethod {
+
+    public static double getZeroPositionsWithPrecisionCondition(
+            Function<Double, Double> function,
+            double leftEndOfCompartment,
+            double rightEndOfCompartment,
+            double epsilon
+    ) throws Exception {
```
Comment (kambed) [inline src/main/java/pl/numerki/backend/SecantMethod.java]: exception

Comment (kambed) [inline src/main/java/pl/numerki/backend/SecantMethod.java]: exception

File: src/main/java/pl/numerki/backend/Bisection.java [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+        for (int i = 0; i < numberOfIterations; i++) {
+            double compartmentMiddle = (leftEndOfCompartment + rightEndOfCompartment) / 2;
+            if (function.apply(compartmentMiddle) == 0) {
+                return compartmentMiddle;
+            }
+            if (function.apply(leftEndOfCompartment) * function.apply(compartmentMiddle) < 0) {
```
Comment (kambed) [inline src/main/java/pl/numerki/backend/Bisection.java]: You can extract this to function and get rid of else

Comment (kambed) [inline src/main/java/pl/numerki/backend/Bisection.java]: You can extract this to function and get rid of else

**Review by kambed — CHANGES_REQUESTED** [review]

File: src/main/java/pl/numerki/frontend/MainFormController.java (hunk 5 above)
Comment (RadekBucki) [reply src/main/java/pl/numerki/frontend/MainFormController.java]: the error is below :), it adjust to error for example only one of zero position calculation method
```
                        "Metoda bisekcji: " + (bisectionException ? "NanN" : resultBisection) +
                        "\nMetoda siecznych: " + (secantException ? "NaN" : resultSecant),
```

**Review by kambed — APPROVED** [review]

**Pull request merged**